File: src/main/java/com/github/hirsivaja/ip/IpHeader.java
package com.github.hirsivaja.ip;

import com.github.hirsivaja.ip.ipv4.Ipv4Header;
import com.github.hirsivaja.ip.ipv6.Ipv6Header;
import java.nio.ByteBuffer;

public sealed interface IpHeader permits Ipv4Header, Ipv6Header {
    void encode(ByteBuffer out);
    byte[] generatePseudoHeader();
    int length();
    int pseudoHeaderLength();
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/header/IpHeader.kt
package com.github.hirsivaja.ip.packet.header

import com.github.hirsivaja.ip.protocol.IpProtocol
import java.nio.ByteBuffer

sealed interface IpHeader {
    fun encode(out: ByteBuffer)
    fun generatePseudoHeader(): ByteArray
    fun length(): Int
    fun pseudoHeaderLength(): Int
    fun protocol(): IpProtocol
}
